lnrpc/routerrpc/router.proto

```diff
@@ -85,6 +89,121 @@ message RouteFeeResponse {
     int64 time_lock_delay = 2;
 }
 
+message SendToRouteRequest {
+    /// The payment hash to use for the HTLC.
+    bytes payment_hash = 1;
+
+    /// Route that should be used to attempt to complete the payment.
+    lnrpc.Route route = 2;
+}
+
+message SendToRouteResponse {
+    /// The preimage obtained by making the payment.
+    bytes preimage = 1;
+
+    /// The failure message in case the payment failed.
+    Failure failure = 2;
+}
+
+message Failure {
+    enum FailureCode {
+        RESERVED = 0;
+        UNKNOWN_PAYMENT_HASH = 1;
+        INCORRECT_PAYMENT_AMOUNT = 2;
+        FINAL_INCORRECT_CLTV_EXPIRY = 3;
+        FINAL_INCORRECT_HTLC_AMOUNT = 4;
+        FINAL_EXPIRY_TOO_SOON = 5;
+        INVALID_REALM = 6;
+        EXPIRY_TOO_SOON = 7;
+        INVALID_ONION_VERSION = 8;
+        INVALID_ONION_HMAC = 9;
+        INVALID_ONION_KEY = 10;
+        AMOUNT_BELOW_MINIMUM = 11;
+        FEE_INSUFFICIENT = 12;
+        INCORRECT_CLTV_EXPIRY = 13;
+        CHANNEL_DISABLED = 14;
+        TEMPORARY_CHANNEL_FAILURE = 15;
+        REQUIRED_NODE_FEATURE_MISSING = 16;
+        REQUIRED_CHANNEL_FEATURE_MISSING = 17;
+        UNKNOWN_NEXT_PEER = 18;
+        TEMPORARY_NODE_FAILURE = 19;
+        PERMANENT_NODE_FAILURE = 20;
+        PERMANENT_CHANNEL_FAILURE = 21;
+    }
+
+    /// Failure code as defined in the Lightning spec
+    FailureCode code = 2;
+
+    /**
+    The node pubkey of the intermediate or final node that generated the failure
+    message.
+    **/
+    bytes failure_source_pubkey = 3;
+
+    /// An optional channel update message.
+    ChannelUpdate channel_update = 4;
+
+    /// A failure type-dependent htlc value.
```

Comment:
Perhaps we can morph these fields into an appropriate `oneof` statement instead? This would make it more explicit when the user should expect these fields to be populated.

Reply:
I tried that first, but decided to keep it simple because `oneof` isn't so nice in some languages. This API is for really advanced users that interpret their own errors, so I am not too worried about interpretation of these fields.